// @gweninterpreter/gwen-web/dist/download.js
"use strict";
/*
 * Copyright 2021 Jacob Juric
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = __importStar(require("fs"));
var path_1 = __importDefault(require("path"));
var os_1 = __importDefault(require("os"));
var axios_1 = __importDefault(require("axios"));
var cachedir_1 = __importDefault(require("cachedir"));
var decompress_1 = __importDefault(require("decompress"));
var progress_1 = __importDefault(require("progress"));
var url_join_1 = __importDefault(require("url-join"));
var files_1 = require("./files");
var storedVersionPath = (0, cachedir_1.default)("gwen-web");
function startDownload(version, mavenRepo) {
    var e_1, _a;
    return __awaiter(this, void 0, void 0, function () {
        var downloadLocation, _b, _c, downloadStream, progress, outputStream, _d, _e, chunk, e_1_1, sha1, etagSum, e_2;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0: return [4 /*yield*/, (0, files_1.fileExists)(path_1.default.join(storedVersionPath, "gwen-web-".concat(version)))];
                case 1:
                    if (_f.sent()) {
                        return [2 /*return*/, {
                                status: "done",
                            }];
                    }
                    console.log("Downloading Gwen-Web v".concat(version, "..."));
                    _f.label = 2;
                case 2:
                    _f.trys.push([2, 18, , 19]);
                    _c = (_b = path_1.default).join;
                    return [4 /*yield*/, fs_1.promises.mkdtemp(path_1.default.join(os_1.default.tmpdir(), "gwen-web-"))];
                case 3:
                    downloadLocation = _c.apply(_b, [_f.sent(), "gwen-web-".concat(version, ".zip")]);
                    return [4 /*yield*/, axios_1.default.get((0, url_join_1.default)(mavenRepo, "/org/gweninterpreter/gwen-web/".concat(version, "/gwen-web-").concat(version, ".zip")), {
                            responseType: "stream",
                        })];
                case 4:
                    downloadStream = _f.sent();
                    progress = new progress_1.default("[:bar] :percent :elapseds", {
                        width: 28,
                        head: ">",
                        total: parseInt(downloadStream.headers["content-length"]),
                    });
                    outputStream = fs_1.default.createWriteStream(downloadLocation);
                    _f.label = 5;
                case 5:
                    _f.trys.push([5, 10, 11, 16]);
                    _d = __asyncValues(downloadStream.data);
                    _f.label = 6;
                case 6: return [4 /*yield*/, _d.next()];
                case 7:
                    if (!(_e = _f.sent(), !_e.done)) return [3 /*break*/, 9];
                    chunk = _e.value;
                    progress.tick(chunk.length);
                    outputStream.write(chunk);
                    _f.label = 8;
                case 8: return [3 /*break*/, 6];
                case 9: return [3 /*break*/, 16];
                case 10:
                    e_1_1 = _f.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 16];
                case 11:
                    _f.trys.push([11, , 14, 15]);
                    if (!(_e && !_e.done && (_a = _d.return))) return [3 /*break*/, 13];
                    return [4 /*yield*/, _a.call(_d)];
                case 12:
                    _f.sent();
                    _f.label = 13;
                case 13: return [3 /*break*/, 15];
                case 14:
                    if (e_1) throw e_1.error;
                    return [7 /*endfinally*/];
                case 15: return [7 /*endfinally*/];
                case 16:
                    console.log("Validating downloaded archive hash...");
                    return [4 /*yield*/, (0, files_1.getFileSha1)(downloadLocation)];
                case 17:
                    sha1 = _f.sent();
                    if (sha1 !== downloadStream.headers["x-checksum-sha1"]) {
                        etagSum = downloadStream.headers["etag"].match(/{SHA1{(.*)}}/);
                        if (etagSum && sha1 !== etagSum[1]) {
                            return [2 /*return*/, {
                                    status: "error",
                                    message: "Failed hash validation! Maybe there was an Internet connection issue. Trying again may resolve the problem.",
                                }];
                        }
                    }
                    return [2 /*return*/, {
                            status: "downloaded",
                            zipPath: downloadLocation,
                        }];
                case 18:
                    e_2 = _f.sent();
                    if (axios_1.default.isAxiosError(e_2) && e_2.response && e_2.response.status === 404) {
                        return [2 /*return*/, {
                                status: "error",
                                message: "Gwen-Web v".concat(version, " doesn't exist. Check your version in package.json and try again."),
                            }];
                    }
                    else if (axios_1.default.isAxiosError(e_2) && (e_2.response || e_2.request || !e_2.code)) {
                        return [2 /*return*/, {
                                status: "error",
                                message: "Failed downloading Gwen-Web v".concat(version, ". Check your internet connection and try again."),
                            }];
                    }
                    else {
                        return [2 /*return*/, {
                                status: "error",
                                message: "An unknown error occured while downloading Gwen-Web.",
                            }];
                    }
                    return [3 /*break*/, 19];
                case 19: return [2 /*return*/];
            }
        });
    });
}
function extractZip(info) {
    return __awaiter(this, void 0, void 0, function () {
        var e_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (info.status !== "downloaded")
                        return [2 /*return*/, info];
                    console.log("Extracting...");
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, (0, decompress_1.default)(info.zipPath, storedVersionPath)];
                case 2:
                    _a.sent();
                    return [2 /*return*/, {
                            status: "done",
                        }];
                case 3:
                    e_3 = _a.sent();
                    return [2 /*return*/, {
                            status: "error",
                            message: "Could not extract Gwen-Web to ".concat(storedVersionPath, "."),
                        }];
                case 4: return [2 /*return*/];
            }
        });
    });
}
function handleError(result) {
    if (result.status === "error") {
        console.log(result.message);
        process.exit(1);
    }
}
function download(version, mavenRepo) {
    return __awaiter(this, void 0, void 0, function () {
        var dlResult, extractResult;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, startDownload(version, mavenRepo)];
                case 1:
                    dlResult = _a.sent();
                    handleError(dlResult);
                    return [4 /*yield*/, extractZip(dlResult)];
                case 2:
                    extractResult = _a.sent();
                    handleError(extractResult);
                    console.log("Downloaded Gwen-Web v".concat(version));
                    return [2 /*return*/];
            }
        });
    });
}
exports.default = download;
